### ISSUE TITLE
Permit [Range] on enums

### ISSUE THREAD

**Issue by jnm2** [nunit/nunit#3292]
This could be useful in a situation where you don't want all enum values, such as `[Range(Keys.D0, Keys.D9)] Keys key`.

The workaround right now would be an int range which the first line of the test could add to enum value at the starting end of the range. Ints in the test name aren't be as nice as enum names would be in the general case. Also, an int range is not very nice if there are gaps in the enum's numeric values.

Rather than incrementing, I'd expect all unique values of the enum between the chosen min and max to be used even if there are gaps. `[Flags]` enums should probably be disallowed for now at least.

Related: https://github.com/nunit/nunit/issues/3246

**Comment by stevenaw** [comment]
I'd be happy to take this on

**Comment by jnm2** [comment]
That would be great! @nunit/framework-team Can you confirm whether this is something we will add?

**Comment by CharliePoole** [comment]
As I already commented in #3246, I don't think a "range" of enums is a thing. Enums have no ordering and are not able to be incremented without conversion. They are more like a set than a sequence.

I don't think this should be a feature at all.

**Comment by jnm2** [comment]
Hmm. I don't want to be difficult, but the assertion that it's not a thing just isn't compelling to me. What term would you suggest for the real thing which is "use all enum members between Foo and Bar"?

C# defines inequality operators (not to mention increment and addition etc), so the idea of a range being a subset of the full set is a coherent concept and one I really could have used last week.

**Comment by jnm2** [comment]
I'm sorry, it was just now sinking in that you didn't think `>=` and `<=` were defined on enums. They are, and I can point to production code that makes use of this. Here's an example of other operations with enums:

```c#
var key = ConsoleKey.D4;
key++;
key += ConsoleKey.NumPad0 - ConsoleKey.D0;
Console.WriteLine(key); // Prints "NumPad5"
```

The only thing I'd rather defer as I mentioned above is how to handle flags enums, since I'm not sure what the expectation would be. But non-flags enum ranges seem straightforward to me.

Given that `>=` and `<=` exist, is there any downside to expanding RangeAttribute to enumerate ranges of sets in addition to ranges of ints?

**Comment by CharliePoole** [comment]
Operators don't matter IME. The intrinsic nature of the type does. A non-flags enum value represents one value, which has no intrinsic relation to any other allowed values. For example, we define five values for ProcessModel. Is Separate intrinsically greater than Single? If so, is it one more, two more or whatever? What does it mean if the order of definition doesn't match the underlying order of ints. What about aliases, like Single and InProcess.

More to the point, what happens if I change the int values? In general, that is not supposed to make any difference with an enum. But it would make a big difference in how this proposed feature would work.

And then this: If I disagree with you, it must mean that I lack some basic knowledge of C#. Really?

FWIW, my opinion is that the C# designers erred in allowing comparisons and probably also in allowing you to specify what constants would be used for each named enum value. They could have allowed initial values based on prior defined names, but not constants. The compiler could pick the values appropriately for all others, even in the case of Flags enums.

**Comment by CharliePoole** [comment]
Ah... one other thing. I completely see how this could be implemented. It would be easy. I thought that was obvious, but maybe I'd better say it given some of your comments.

The question is __should__ we implement it.

**Comment by jnm2** [comment]
The intrinsic nature of many enums gives ranges a coherent meaning. Keys, windows messages, log levels, etc. If you can point to an enum where a range is not useful, why should that stop someone from using a range where it is useful?

The fact that enums have a strong ordering means that ranges are intrinsic to their very nature.

> And then this: If I disagree with you, it must mean that I lack some basic knowledge of C#. Really?

Not at all! Please give me credit for my attempt to really hear you and not talk past you by making assumptions. You said, "Enums have no ordering and are not able to be incremented without conversion." As a matter of fact, enums have a well-defined ordering and they are able to be incremental without conversion. This has nothing to do with whether I disagree with you.

**Comment by jnm2** [comment]
The design of many enums gives ranges a coherent meaning. Keys, windows messages, log levels, etc. If you can point to an enum where a range is not useful, why should that stop someone from using a range where it is useful?

The fact that enums have a strong ordering means that ranges are intrinsic to their very nature.

> What about aliases, like Single and InProcess.

You can't pass an enum member in .NET, only an enum value. Since this proposal is all about generating enum values, aliases aren't something we have to deal with.

> And then this: If I disagree with you, it must mean that I lack some basic knowledge of C#. Really?

Not at all! This comment troubles me. This is where I notice that the conversational has turned personal, and it seems unjust to me. Please give me credit for trying to really hear you and not talk past you by making assumptions. You said, "Enums have no ordering and are not able to be incremented without conversion." As a matter of fact, enums have a well-defined ordering and they are able to be incremented without conversion. This has nothing to do with whether I disagree with you. I avoided directly contradicting you out of good will, but then was worried that by doing so I was obscuring the underlying issue and that we would be talking past each other which only leads to frustration.

I'm doing my best in a situation that is hard for me to parse. I always hope you'll give me the benefit of the doubt as I do the same for you. I see true value in this issue and it would be disingenuous for me to say otherwise or cease responding. But does it have to be so emotionally draining? I can't do this.

**Comment by CharliePoole** [comment]
You asked for team input as to whether we should add this.

**Comment by CharliePoole** [comment]
It's hard to reply to messages that keep changing via GitHub edit.

**Comment by jnm2** [comment]
Yes, I asked for input! And you said no on the basis that enums are intrinsically not something that can have ranges. In your words, "it's not a thing." But it is a thing and I can demonstrate it, and at least two people have seen the real-world use and asked for it. If you agree that what we have asked for really is a thing, which I believe is a true and worthwhile point, we can move on to any other considerations. After that, the outcome may be "no, we don't think this addition would make NUnit better," and that would be fine. My assumption is that you (like me) would be glad to notice when the basis of a decision is not the whole picture.

**Comment by CharliePoole** [comment]
@jnm2 You wrote me offline and I'll reply that way, but I want to respond here to something that you said publicly...

>> And then this: If I disagree with you, it must mean that I lack some basic knowledge of C#. Really?

> Not at all! This comment troubles me. This is where I notice that the conversational has turned personal, and it seems unjust to me. 

Yes, it did turn personal, at least in the sense that I felt I knew you well enough to make a joke at which (I thought) you wouldn't take offense. Clearly I don't, and I apologize for overstepping.

If I had been seriously offended by your comments about the C# language, I hope you know me well enough to realize that I'd most likely say something like "I'm offended by that" and would most assuredly not make a joke out of it.

OTOH maybe my sense of humor is weird and nobody else sees that "Really?" makes it funny. :confused:

**Comment by CharliePoole** [comment]
You're telling me that I have to be extremely careful how I word my opinions. That's a useful takeaway for me personally but might not be so good for the team as a whole.

Yes, there are many other considerations we should look at like "Is this a good idea for NUnit as a whole?" We haven't gone there because we are stuck on this single point. You seem to believe we're stuck because I won't surrender my view that this is the wrong way to use enums. I think we are stuck because you insist on arguing the various points sequentially, rather than collecting everyone's views, having a discussion and seeing if a consensus arises. IME consensus never rises out of argument.

Is that a fair picture of where we are?

**Comment by ChrisMaddock** [comment]
I like this feature, and it seems harmless to me. I agree that .NET enums have an intrinsic order, built in by the compiler, if not specified explicitly in code. This may not be relevant to all defined enums (e.g. `ProcessModel`, as you say, Charlie), but regardless those enums still have an order - and it's for the user to decide if that order has any meaningful value.

A nice usecase from my particular line of work might be days of the week, e.g. `[Range(DayOfWeek.Monday, DayOfWeek.Friday)]`.

@rprouse - I think this one might value your thoughts, if @stevenaw is still happy to contribute this feature? 🙂

**Comment by CharliePoole** [comment]
@ChrisMaddock That's a bad or point of view, i think. I'd prefer to day that the concept of an enum doesn't have an intrinsic order, but the .Net implementation does, but for most people that may be splitting hairs. I'd say no, because o prefer not to encourage people to this of rooms that way, but apparently some already do, due to the fact that you can assign specific values.

**Comment by oznetmaster** [comment]
I agree totally with CharliePool (who has disappeared from the list of members).

I suppose that when a value is explicitly assigned to all enum members, then there is an explicit order.  However, I have seen many enums where some values are assigned, and some are not.  What does that mean for the order?  Or if multiple enum members are assigned the same value, which also happens often.  Or when some enum member values are expressions of others in the enum (enum | enum | enum).

I really have to agree that an enum (except in some very special cases) does not have an intrinsic order to which the Range attribute should apply.

**Comment by CharliePoole** [comment]
@oznetmaster I'm still here. :smiley: Sometimes the @ prompts don't seem to work.

**Comment by CharliePoole** [comment]
Aside from a philosophical difference about enums, I do see this as potentially opening a can of worms... For example, some people may not understand how values are assigned implicitly, particularly if explicit assignments are included. Others may use it on enums without a natural sequencing. Anyone who misuses it will probably file a bug report when it doesn't work as expected. The folks who truly understand the limitations already know how to do this where needed at the cost of some extra typing. OTOH, it's not a line in the sand for me if somebody wants to provide it.

**Comment by jnm2** [comment]
The introductions to C# enums that I've come across are heavily number-centric. https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/enum appears the same way. If someone is not getting the result they expected, my guess is that the numeric values would be the first thing they would look at. Worst case, we can point folks to the C# docs.

One thing I suggested in the original proposal is disallowing flags enums for now until and unless the motivating use cases come to light. This would be through a runtime test error and our Roslyn analyzer would also want to provide an error diagnostic in the source as the user types. This way, we don't guess wrong and paint ourselves into a compatibility corner.

To repeat a bit about aliases (two enum members with the same _value_) which is buried above: There isn't a way in .NET to pass a reference to a particular enum member directly, since an enum instance comprises only the distinct integral value. That means we can happily ignore aliases because we are generating enum _values_.

**Comment by jnm2** [comment]
For reference, this is what the extra typing looks like to achieve the goal of useful test names or skipping gaps in the sequence:

```cs
public static IEnumerable<DayOfWeek> Weekdays =>
    from DayOfWeek value in Enum.GetValues(typeof(DayOfWeek))
    where value >= DayOfWeek.Monday && value <= DayOfWeek.Friday
    select value;

[TestCaseSource(nameof(Weekdays))]
public void Test(DayOfWeek dayOfWeek)
{
    // ...
}
```

Not terrible by any means, but it feels the same to me as it would feel to have to do this with ints... like the [Range] attribute should be able to do it.

**Comment by oznetmaster** [comment]
If an enum with aliases is enumerated, it raises a question.  Should the aliased value be returned for each alias?

I have seen many posts about that being a desired result in some cases.

Should that be a consideration here?

**Comment by jnm2** [comment]
@oznetmaster That would result in two identical test cases running. I haven't thought of a way that it could be useful to run the same test method with the same value twice. The test name would be the same too unless we add a way to pass additional context to the test name generator beyond the argument values themselves.

**Comment by oznetmaster** [comment]
How about allowing the range attribute on an enum with a string parameter instead of an int.  This would give a set of the names of the enum members.

That would allow a more generic case, if desired.  The test could then parse that string to get the value, or use it in some other way if desired.

**Comment by oznetmaster** [comment]
With a min/max parameter, those values could perhaps delimit the set of names returned (by their values).

**Comment by jnm2** [comment]
@oznetmaster That seems pretty hypothetical to me. It's not how enums are thought of in .NET code. Let's not do that unless someone asks for it. Even then, I'd say they are working directly against the meaning of enums since aliases are the way to say there is no difference. What you're talking about is Enum.GetNames where you enumerate enum members rather than enum values. I'd prefer to keep it simple and support only passing enum values at least as far as this proposal goes.
